### ISSUE TITLE
Framework Support

### ISSUE THREAD

**Issue by s1rc** [ankitpokhrel/tus-php#41]
Is it possible for this to run on request within a framework, like Laravel?

I want to test this out, however with deployment it would be ideal to have requests handled by our app instead of running a separate server.

**Comment by ankitpokhrel** [comment]
@s1rc I added a rough doc for Laravel integration: https://github.com/ankitpokhrel/tus-php/wiki/Laravel-Lumen-Integration

**Comment by s1rc** [comment]
Thank you, I will give this a try!

**Comment by s1rc** [comment]
So far so good!

One quick change to your Wiki article. In the route definition, the parameter needs to be optional.

```php
Route::any('/tus/{any?}', function () {
    $response = app('tus-server')->serve();

    return $response->send();
})->where('any', '.*');
```

**Comment by ankitpokhrel** [comment]
Yes true, thanks :)

**Comment by s1rc** [comment]
Not sure if I should open a new issue, but do have some follow up questions.

1. Where are the files stored during upload? In the redis cache?
2. If using redis cache I cannot resume, if using file cache upload finishes 100% immediately. (both using Uppy)
3. Is there a completion handler for the server? I want to perform some post upload actions (i.e. add to a media library and remove from the tus upload directory)

**Comment by ankitpokhrel** [comment]
1. Files are stored in the location you specify while setting up the server. By default, it uses `uploads` folder in the installation directory.
    ```
    ->setUploadDir(storage_path('app/public/uploads')); // upload dir
    ```

2. Make sure that `tus-server` can connect the redis server. You can use `REDIS_HOST` and `REDIS_PORT` env to pass redis host and port. You can also directly create and pass redis object.
    ```
    // TusServiceProvider.php

    $redis = new TusPhp\Cache\RedisStore([
        'host' => '127.0.0.1',
        'port' => '6379',
    ]);

    $server = new Server($redis);
    ```
    > if using file cache upload finishes 100% immediately

    I guess you are testing in your local setup. This is related to uppy and the reason is that the progress bar is updated as soon as the requested byte is queued to sent to the server. This process is too fast in local and it reaches 100% immediately. If you use actual server this won't happen.

3. No there isn't but if you are using uppy you can use [`upload-success` event](https://uppy.io/docs/uppy/) for post upload operations.

**Comment by s1rc** [comment]
1. I have it set to

**Comment by ankitpokhrel** [comment]
Hi @s1rc,

I just tried in one of my Laravel's installation and it seems to be working fine. I hope you properly setup your tus endpoint for uppy.
```js
const uppy = Uppy.Core({debug: true, autoProceed: false})
    .use(Uppy.Dashboard, {target: '#uppyUploader', inline: true})
    .use(Uppy.Tus, {endpoint: 'http://your.app/tus'}) // tus endpoint
    .run()
```

Note that if you upload same file again while was successfully uploaded previously it won't get uploaded unless you clear your redis or file cache. File cache is saved in `.cache/tus_php.cache`.

Flush your cache first and then try uploading some large files and pause it after a while even if uppy shows 100%. You should see partially uploaded file in your uploads dir. Resume it again and when the upload completes, you should be able to see the complete file.

**Comment by s1rc** [comment]
I set it up exactly as you wrote the Wiki. 

I've tried file cache and deleting `vendor/ankitpokhrel/tus-php/.cache/tus_php.cache`, as well as deleting all records from redis when using redis config.

During upload there are no partially uploaded files in the folder defined at `setUploadDir(storage_path('tus'))`

I've tried setting the Laravel environment cache to file and redis, no change.

You can see here that pausing and resuming the file just starts over
[https://file.io/HcBhXx](url)

And when a file is at 100% there is a long pause while the file is copied into the folder defined at ``setUploadDir(storage_path('tus'))`
[https://file.io/mqFjkX](url)

**Comment by ankitpokhrel** [comment]
Hi @s1rc,

The progress bar is updated by uppy as soon as the first byte is sent to the server and don't wait for the actual byte to be updated. This process is very quick in local so the progress bar reaches 100% immediately. This won't happen in actual server as there will be lag when sending first byte to the server.

> You can see here that pausing and resuming the file just starts over

I see that the file is 2.5gb which will take some time to upload (even partially). Try by uploading small file lets say 50 mb. When the progress bar reaches 100%, wait for 2-3 seconds and pause it, there should be partially uploaded file in your uploads dir.

> And when a file is at 100% there is a long pause while the file is copied into the folder

There is long pause because file is now actually being uploaded to the server. Again, this happens only in local because of the way uppy progress bar works.

If you have a test server, I recommend you to check on that and things will be clear for you.

I know this is confusing but it happens only in local and I haven't found a way to fixed that yet. If you can figure out how to fix it that would be a great help. I will try to check with uppy team later.

**Comment by s1rc** [comment]
@ankitpokhrel thanks for the feedback.

I tested it on a development server on AWS and the behaviour seems the similar. Upload will not resume but instead starts over. When the upload is complete the Tus server can't seem to read the file from the specified directory. (I checked and the file does not exist in the upload directory)

**Comment by ankitpokhrel** [comment]
I will try to replicate the issue as soon as I get some time and will get back to you

**Comment by ankitpokhrel** [comment]
Closed in favor of #44
